File: C++/0143-reorder-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    void reorderList(ListNode* head) {

        if (head==NULL || head->next == NULL)
            return;
        ListNode* prev = NULL;
        auto slow = head;
        auto fast = head;

        while (fast && fast->next){
            prev = slow;
            slow = slow->next;
            fast = fast->next->next;
        }

        prev->next = nullptr;

        auto l1 = head;
        auto l2 = reverseList(slow);

        ListNode dummy {0};
        auto current = &dummy;
        while(l1 && l2){
            auto l1Next = l1->next;
            auto l2Next = l2->next;

            current->next = l1;
            current = l1;
            l1 = l1->next;

            current->next = l2;
            current = l2;
            l2 = l2->next;
        }
    }
private:
    ListNode* reverseList(ListNode* head){
        auto current=head;
        ListNode *prev = NULL;

        while(current){
            auto next = current->next;
            current->next = prev;
            prev = current;
            current = next;
        }
        return prev;
    }
};



/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    void reorderList(ListNode* head) {

        if (head==NULL || head->next == NULL)
            return;
        ListNode* prev = NULL;
        auto slow = head;
        auto fast = head;

        while (fast && fast->next){
            prev = slow;
            slow = slow->next;
            fast = fast->next->next;
        }

        prev->next = nullptr;

        auto l1 = head;
        auto l2 = reverseList(slow);

        //ListNode dummy {0};
        //auto current = &dummy;
        while(l1){
            ListNode *l1Next = l1->next;
            ListNode *l2Next = l2->next;
            l1->next = l2;
            if(!l1Next)
                break;
            l2->next = l1Next;
            l1 = l1Next;
            l2 = l2Next;
        }
    }
private:
    ListNode* reverseList(ListNode* head){
        auto current=head;
        ListNode *prev = NULL;

        while(current){
            auto next = current->next;
            current->next = prev;
            prev = current;
            current = next;
        }
        return prev;
    }
};

File: Python/0143-reorder-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reorderList(self, head: ListNode) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        if head == None or head.next == None:
            return


        def reverseList(head):
            prev = None
            curr = head

            while curr:
                nextNode = curr.next
                curr.next = prev
                prev = curr
                curr = nextNode

            return prev



        def merge(l1, l2):
            dummy = ListNode(0)
            current = dummy
            while l1 and l2:
                current.next, current, l1 = l1, l1, l1.next
                current.next, current, l2 = l2, l2, l2.next

#             while(l1):
#                 l1Next = l1.next
#                 l2Next = l2.next

#                 l1.next = l2

#                 # if length is odd, l1 will be shorter. Otherwise if the length is even, still this condition works.
#                 if l1Next == None:
#                     break

#                 l2.next = l1Next
#                 l1 = l1.next
#                 l2 = l2.next



        slow, prev, fast = head, None, head
        # getting to the middle of list
        while fast and fast.next:
            prev = slow
            slow = slow.next
            fast = fast.next.next

        prev.next = None
        # l1 is the head of first half list
        # l2 is head of second half list reversed
        l1, l2 = head, reverseList(slow)

        merge(l1,l2)

        return head






# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reorderList(self, head: ListNode) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        if head == None or head.next == None:
            return


        def reverseList(head):
            prev = None
            curr = head

            while curr:
                nextNode = curr.next
                curr.next = prev
                prev = curr
                curr = nextNode

            return prev



        slow, prev, fast = head, None, head
        # getting to the middle of list
        while fast and fast.next:
            prev = slow
            slow = slow.next
            fast = fast.next.next

        prev.next = None
        # l1 is the head of first half list
        # l2 is head of second half list reversed
        l1, l2 = head, reverseList(slow)

        while l1:
            l1Next, l2Next = l1.next, l2.next
            l1.next = l2
            if not l1Next:
                break
            l2.next = l1Next
            l1, l2 = l1Next, l2Next


        return head
